Tile- and Animation-specific hitboxes and collision rework

michelfinley opened the pull request:
## Summary

This PR implements hitboxes being able to defined within Tiled's tileset editor, entity hitboxes being able to be defined for each individual animation frame. It also features a small rework of the collision system that might make it feel a bit more intuitive, as well as fix a bug of an entity on-collision always being pushed against the direction it is looking in, regardless of the direction it has or should have collided with the opposing object.

## Checklist

- [X] I have tested this change locally and it works as expected.
- [X] I have made sure that the code follows the formatting and style guidelines of the project.
- [X] Since this PR modifies related documentation, I have proofread files for spelling and grammar errors.

## Labels
`game-art` `game-gameplay`

michelfinley committed:
Object Hitbox Test

level.py
- pass object hitboxes to collidable objects
- add debugging overlay displaying the hitboxes of all entities

main.tmx
- correct dimensions of collision rectangles

objects.tsx
- add rectangular hitboxes to all currently used objects

sprites.py
- remove parameter shrink from CollideableSprite
- change the Plant parent class to Sprite
- add hitbox parameter to Tree, NPC and Player, that transforms hitbox_rect to the dimensions of the given relative hitbox
- update Entity.collision to work with all CollideableSprites
- fix Entity.add_resource to avoid a game crash when trying to chop a tree

support.py
- add get_object_hitboxes which reads the given tileset and returns a dict with all tile hitboxes

see also https://github.com/sloukit/pydew-valley-uzh/issues/35

michelfinley committed:
Enable NPCs by default + improve visual hitbox preview

michelfinley committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into test-object-collision

# Conflicts:
#	src/level.py
#	src/sprites.py
#	src/support.py

michelfinley committed:
Reimplement hitbox test

michelfinley committed:
Restructured hitbox initialisation

Collideable objects should now be initialised as sprites.base.CollideableMapObject using one of the MapObjectTypes from Level.map_objects

michelfinley committed:
Made CollideableSprite an abstract class, remove redundant attribute Entity.plant_collide_rect

michelfinley committed:
feat(entity): adjusted collision system

sprites/base.py
- the type of CollideableSprite.hitbox_rect has been changed to pygame.FRect for smoother animations

sprites/entity.py
- now requires an EntityAsset as attribute instead of frames
- reworked the collision method:
Entities no longer collide in a specific direction, but instead get put at the nearest edge of the object they collided with

enums.py
- added Direction and EntityState enum to simplify code maintenance

screens/level.py
- adjustments for reworked asset loading

mapobjects.py
- changed hitbox type to pygame.FRect

npc/npc.py + sprites/player.py
- adjustments for reworked asset loading and collision system

sprites/setup.py
- new file to make individual hitboxes for each animation frame possible
- the assets for individual entities will now get set up in EntityAssets.setup

michelfinley committed:
refactor: fix lint issues, remove duplicates, add comments and extract test function

- extracted logic regarding the visual display of hitboxes from Level.draw

michelfinley committed:
Merge remote-tracking branch 'refs/remotes/upstream-raphael/hitboxes' into test-object-collision

# Conflicts:
#	data/maps/main.tmx
#	src/npc/npc.py
#	src/screens/level.py
#	src/sprites/base.py
#	src/sprites/entities/entity.py
#	src/sprites/entities/player.py
#	src/sprites/objects/plant.py
#	src/sprites/objects/tree.py

michelfinley committed:
fix: merge issues and normalisation of entity directions

- add attribute is_colliding to Entity and rename Entity.collision to Entity.check_collision again

michelfinley committed:
docs: update CONTRIBUTING.md with a new guide on how to add custom collisions

raphael-pietrzak commented on src/mapobjects.py:
```diff
@@ -0,0 +1,83 @@
+                ).convert_alpha()
+
+                objectgroup = child.find("objectgroup")
+                if objectgroup is not None:
```
you can reduce it like if objectgroup:

michelfinley replied:
Will do 😄

raphael-pietrzak commented on src/mapobjects.py:
```diff
@@ -0,0 +1,83 @@
+                objectgroup = child.find("objectgroup")
+                if objectgroup is not None:
+                    hitbox = objectgroup.find("object")
+                    if hitbox is not None:
```
same as before

raphael-pietrzak commented on src/mapobjects.py:
```diff
@@ -0,0 +1,83 @@
+        self.objects = get_object_hitboxes(tileset.value)
+
+
+def get_object_hitboxes(path_to_tileset: str) -> dict[int, MapObjectType]:
+    """
+    Returns a dictionary mapping the IDs of all tiles in a given tileset
+    to the first hitbox created with the Tiled Tile Collision Editor.
+    If no hitbox is found, it uses the dimensions of the Tile's image.
+
+    This function is currently experimental and has only been tested with rectangular hitboxes.
+    It is also currently limited to integer values within the dimensions of the hitbox.
+
+    :return: Dictionary mapping the IDs of all tiles in a given tileset to its hitbox (x, y, width, height)
+    """
+    tree = ElementTree.parse(path_to_tileset)
+    root = tree.getroot()
+
+    objects = {}
+    try:
+        for child in root:
+            obj = {
+                "id": None,
+                "hitbox": None,
+                "image": None
+            }
+            if child.tag == "tile":
+                obj["id"] = child.attrib.get("id")
+
+                image = child.find("image")
+                if image is None:
+                    raise Exception(f"Malformed tileset {path_to_tileset}\n")
+
+                obj["image"] = pygame.image.load(
+                    os.path.join(os.path.dirname(path_to_tileset),
+                                 image.attrib.get("source"))
+                ).convert_alpha()
+
```
please do multiple methods for this 60 line method

michelfinley replied:
Oh yeah, that's probably for the best 😅

raphael-pietrzak commented on src/npc/npc.py:
```diff
@@ -45,16 +45,22 @@ def __init__(
             apply_tool
         )
 
+        first_frame_hitbox = (self.assets[self.state][self.facing_direction]
```
a method should be existing self.assets.get_hitbox()

raphael-pietrzak replied:
what's the utility of first frame hitbox ?

michelfinley replied:
It's just the hitbox that the entity initialises with, though it is probably not necessary. We could just initialise the NPC.hitbox_rect with no arguments and then "initialise" it with the first NPC.update call

raphael-pietrzak commented on src/npc/npc.py:
```diff
@@ -45,16 +45,22 @@ def __init__(
+        first_frame_hitbox = (self.assets[self.state][self.facing_direction]
+                              .get_hitbox(self.frame_index))
+
+        self.hitbox_rect = pygame.FRect(
+            self.rect.left + first_frame_hitbox.x,
+            self.rect.top + first_frame_hitbox.y,
+            first_frame_hitbox.w, first_frame_hitbox.h
+        )
```
this is no needed because the hitbox position will be updated at the first update, also i think a rect is better than a Frect to avoid collisions bugs, hard to explain here

raphael-pietrzak replied:
finally i think we should have self.hitbox_rect = assets.get_hitbox()

michelfinley replied:
You're right, I'll do that right away

raphael-pietrzak commented on src/npc/npc.py:
```diff
@@ -128,8 +134,16 @@ def create_path_to_tile(self, coord: tuple[int, int]) -> bool:
 
-
     def move(self, dt):
+        current_hitbox = (self.assets[self.state][self.facing_direction]
+                          .get_hitbox(int(self.frame_index)))
+
+        self.hitbox_rect.update(
+            (self.rect.x + current_hitbox.x,
+             self.rect.y + current_hitbox.y),
+            current_hitbox.size
+        )
+
```
same

raphael-pietrzak commented on src/npc/npc.py:
```diff
@@ -150,25 +164,28 @@ def update_direction(self):
-        self.direction.x = round(dx / magnitude)
-        self.direction.y = round(dy / magnitude)
-    
+        magnitude = dx + dy
+        if magnitude:
+            self.direction.x = round(dx / magnitude)
+            self.direction.y = round(dy / magnitude)
+        else:
+            self.direction.xy = (0, 0)
+
```
the next path point function already update direction

michelfinley replied:
Oh, this code section is only there to avoid zero division errors. I think it could be replaced by `self.direction = (dx, dy)` and then
```
if self.direction:
   self.direction = self.direction.normalize()
```
just as it is already in `Player.input`

raphael-pietrzak commented on src/screens/level.py:
```diff
@@ -161,28 +165,27 @@ def pf_matrix_setup_collision(self, pos: tuple[float, float], size: tuple[float,
-            Bed(pos, image, (self.all_sprites, self.collision_sprites))
-        elif obj.name == 'Trader':
-            Trader(pos, image, (self.all_sprites, self.collision_sprites))
+            Tree(pos,
+                 self.map_objects.objects[obj.properties.get("id")],
+                 (self.all_sprites, self.collision_sprites, self.tree_sprites),
+                 obj.name, apple_frames, stump_frames)
```
add intermediary variables image and groups,

raphael-pietrzak commented on src/screens/level.py:
```diff
@@ -161,28 +165,27 @@ def pf_matrix_setup_collision(self, pos: tuple[float, float], size: tuple[float,
-            Rock(pos, image, (self.all_sprites, self.collision_sprites))
-        elif obj.name == 'Bed':
-            Bed(pos, image, (self.all_sprites, self.collision_sprites))
-        elif obj.name == 'Trader':
-            Trader(pos, image, (self.all_sprites, self.collision_sprites))
+            Tree(pos,
+                 self.map_objects.objects[obj.properties.get("id")],
+                 (self.all_sprites, self.collision_sprites, self.tree_sprites),
+                 obj.name, apple_frames, stump_frames)
```
so for you all objects are tree_sprites

michelfinley replied:
No, of course not! 😂
I just thought that limiting the amount of graphical objects on the map to a bare minimum in code would be more efficient, so I replaced all Rocks, Beds, and Traders with instances of CollideableMapObject's with the same attributes

raphael-pietrzak commented on src/sprites/entities/entity.py:
```diff
@@ -69,72 +65,90 @@ def __init__(
-                        self.hitbox_rect.bottom = sprite.hitbox_rect.top
-
-        
-        self.rect.midbottom = self.hitbox_rect.midbottom + self.hitbox_offset
-
-        
+            if sprite is not self:
+
+                # Entities should collide with their hitbox_rects to make them able to approach
+                #  each other further than the empty space on their sprite images would allow
+                if isinstance(sprite, CollideableSprite):
+                    if sprite.hitbox_rect.colliderect(self.hitbox_rect):
+                        colliding_rect = sprite.hitbox_rect
+                elif sprite.rect.colliderect(self.hitbox_rect):
+                    colliding_rect = sprite.rect
+
+                if colliding_rect:
+                    distances = (
+                        abs(self.hitbox_rect.right - colliding_rect.left),
+                        abs(self.hitbox_rect.left - colliding_rect.right),
+                        abs(self.hitbox_rect.bottom - colliding_rect.top),
+                        abs(self.hitbox_rect.top - colliding_rect.bottom)
+                    )
+                    shortest_distance = min(distances)
+                    if shortest_distance == distances[0]:
+                        self.hitbox_rect.right = colliding_rect.left
+                    elif shortest_distance == distances[1]:
+                        self.hitbox_rect.left = colliding_rect.right
+                    elif shortest_distance == distances[2]:
+                        self.hitbox_rect.bottom = colliding_rect.top
+                    elif shortest_distance == distances[3]:
+                        self.hitbox_rect.top = colliding_rect.bottom
+
+        self.is_colliding = bool(colliding_rect)
```
i can't let you add this because first all sprites have a hitbox now so we can remove the isinstance.
also it doesn't make sense to move the hitbox instead of the rect (or the position) also a new method to get player side is needed

raphael-pietrzak commented on src/sprites/entities/entity.py:
```diff
@@ -69,72 +65,90 @@ def __init__(
         if not self.tool_active:
-            self.image = current_animation[int(
-                self.frame_index) % len(current_animation)]
+            self.image = current_animation.get_frame(self.frame_index)
         else:
-            tool_animation = self.frames[self.available_tools[self.tool_index]
+            tool_animation = self.assets[
+                EntityState(self.available_tools[self.tool_index])
             ][self.facing_direction]
             if self.frame_index < len(tool_animation):
-                self.image = tool_animation[min(
-                    (round(self.frame_index), len(tool_animation) - 1))]
-                if round(self.frame_index) == len(tool_animation) - \
-                        1 and not self.just_used_tool:
+                self.image = tool_animation.get_frame(
+                    min(round(self.frame_index), len(tool_animation) - 1)
+                )
+                if (round(self.frame_index) == len(tool_animation) - 1
+                        and not self.just_used_tool):
                     self.just_used_tool = True
                     self.use_tool(ItemToUse.REGULAR_TOOL)
```
i have refactored animation method this will be more easier

raphael-pietrzak commented on src/sprites/entities/player.py:
```diff
@@ -34,13 +34,21 @@ def __init__(
             apply_tool
         )
 
+        first_frame_hitbox = (self.assets[self.state][self.facing_direction]
+                              .get_hitbox(self.frame_index))
+
+        self.hitbox_rect = pygame.FRect(
+            self.rect.left + first_frame_hitbox.x,
+            self.rect.top + first_frame_hitbox.y,
+            first_frame_hitbox.w, first_frame_hitbox.h
+        )
+
```
same as before

raphael-pietrzak commented on src/sprites/entities/player.py:
```diff
@@ -122,7 +130,8 @@ def input(self):
             self.direction.x = int(self.controls['right']) - int(self.controls['left'])
             self.direction.y = int(self.controls['down']) - int(self.controls['up'])
-            # self.direction = self.direction.normalize() if self.direction else self.direction
+            if self.direction:
+                self.direction = self.direction.normalize()
```
i prefer when the player can move up and right at the same time this is why i removed normalization

michelfinley replied:
When I tested your implementation, I noticed that the lack of normalization allows the player to move much faster than I thought was intended. With normalisation the player moves ~0.7 * speed in each direction, and without it the player moves 0,3 * faster in each direction, and doubles its total speed

raphael-pietrzak replied:
so we should lower the speed ? because this is really frustrating having one direction at the time

michelfinley replied:
No, not lower the speed but ensure that the player's speed is always limited to `Entity.speed`.
The Player would still move faster if he moves diagonally instead of in two straight lines to the same position, just not at twice the pace of its speed attribute.

raphael-pietrzak commented on src/sprites/entities/entity.py:
```diff
@@ -69,72 +65,90 @@ def __init__(
+                    colliding_rect = sprite.rect
+
+                if colliding_rect:
+                    distances = (
+                        abs(self.hitbox_rect.right - colliding_rect.left),
+                        abs(self.hitbox_rect.left - colliding_rect.right),
+                        abs(self.hitbox_rect.bottom - colliding_rect.top),
+                        abs(self.hitbox_rect.top - colliding_rect.bottom)
+                    )
+                    shortest_distance = min(distances)
```
i think this is bad implementation because the hitbox can have rect size instead of square size so you can have a shortest distance and not colliding in the good direction

michelfinley replied:
Finally fixed 😓

raphael-pietrzak commented on src/sprites/entities/player.py:
```diff
@@ -150,6 +159,24 @@ def input(self):
+
+        self.hitbox_rect.x += self.direction.x * self.speed * dt
+        self.hitbox_rect.y += self.direction.y * self.speed * dt
+        self.check_collision()
```
we need to check collision after each axes moving to know if the player have to be moved vertically or horizontally

michelfinley replied:
I think this is no longer necessary because the function itself now calculates the direction in which the collision took place

raphael-pietrzak commented on src/sprites/entities/player.py:
```diff
@@ -150,6 +159,24 @@ def input(self):
+        self.hitbox_rect.x += self.direction.x * self.speed * dt
+        self.hitbox_rect.y += self.direction.y * self.speed * dt
+        self.check_collision()
+
+        self.rect.update(
+            (self.hitbox_rect.x - current_hitbox.x,
+             self.hitbox_rect.y - current_hitbox.y), self.rect.size)
+
```
same as before we should moving the rect and adjust hitbox, not the opposite

michelfinley replied:
But we only know exactly where we have to place the Rect after calling check_collision, because the collision has to check the updated hitbox_rect and moves it accordingly, right?

raphael-pietrzak replied:
the right way should be :
move the rect
adjust hitbox 
block hitbox (check collision)
adjust rect on hitbox

this doesn't change the gameplay but this looks more logical to me, technically this doesn't matter, i just don't wanna see other classes moving hitbox in purpose to move the player this is unreadable, other people will not understand

michelfinley replied:
It may seem more logical, but I suspect this might actually have a small impact on the performance. Because in the end you would still set the position of the Rect one additional time.

raphael-pietrzak commented on src/sprites/setup.py:
```diff
@@ -0,0 +1,171 @@
+    elif isinstance(item, pygame.Rect):
+        item.update(
+            item.left * factor,
+            item.top * factor,
+            item.width * factor,
+            item.height * factor
+        )
+
+
+_EntityHitboxTypes = (
+        dict[EntityState, dict[Direction, list[pygame.Rect] | pygame.Rect]]
+        | dict[Direction, list[pygame.Rect] | pygame.Rect]
+        | pygame.Rect
+)
+
+
+def entity_importer(
+        path: str, size: int, directions: list[Direction],
+        hitboxes: _EntityHitboxTypes
+) -> dict[EntityState, dict[Direction, _AniFrames]]:
+    scale_nested_rects(hitboxes, SCALE_FACTOR)
+    states = {}
+    for folder_path, sub_folders, file_names in os.walk(path):
+        for file_name in file_names:
+            current_state = EntityState(file_name.split(".")[0])
+            if isinstance(hitboxes, pygame.Rect):
+                states[current_state] = state_importer(
+                    os.path.join(folder_path, file_name),
+                    size=size,
+                    directions=directions,
+                    hitboxes=hitboxes,
+                )
+            else:
+                states[current_state] = state_importer(
+                    os.path.join(folder_path, file_name),
+                    size=size,
+                    directions=directions,
+                    hitboxes=(hitboxes.get(current_state) or hitboxes),
+                )
+    return states
```
I don’t understand why no one codes like @clear-code-projects; their way of coding is super clear with only self attributes, no static methods, no class variables. However, it stays hidden in a file, which I’m not a huge fan of

michelfinley replied:
I only used these two variables to hint a type for readability because it would be way too long to just put it in entity_importer.
Also, for any future to these two importer functions, I thought it would be good if they both knew what type of argument they needed, and they are protected variables so they only stay within that particular file.
At the moment, however, we only have the Rabbit sprite that is loaded with multiple hitboxes, so I could adjust the parameter type to be exactly that that the setup method currently uses, if that's more legible.

raphael-pietrzak replied:
this is not what i meant to say, in @clear-code-projects, the structure is way more clean like this :


class Class:
    def __init__(self, pos, image, rect):       <-- no types (just clear var names)
         self.pos = pos                                     <-- self attributes
         self.image = image
         self.rect = rect
         
     def import_entities(self, path, hitboxes):      <-- methods name are actions
         ...                                                                   <-- no isinstance calls

michelfinley replied:
solved 😄

michelfinley committed:
enha / refactor: new entity properties and attributes, split mapobjects._get_object_hitboxes into two functions

- Entity.state, facing_direction and facing_index have been converted to properties, so that they automatically update the new Entity._current_frame and _current_hitbox attributes

- mapobjects._get_obj_hitbox now uses an instance of MapObjectType directly instead of a dict

michelfinley committed:
Merge remote-tracking branch 'refs/remotes/upstream-raphael/hitboxes' into test-object-collision

michelfinley committed:
refactor(entity): Reworked the Entity.check_collision method (again)

- When colliding with an object, Entities now use their hitbox from the last frame to check where they did collide first

michelfinley committed:
refactor(entity): animation + state rework, simplify hitbox setup

- the Entity.state does now correctly reflect what state the Entity is in, at all times

- simplified Entity.animate to use the new Entity attributes

- Sprite hitboxes are now setup using a setup._Hitbox object, which simplifies the process of setting up custom hitboxes and improves the code readability

- removed FIXME comment fixed in previous commit
